### PR TITLE
change ResultRef.ResultsIndex from int to *int

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes
This commit closes 7392. When we introduced array results, we added validation funciton to check if the result reference is out of the array bound, in the cases of refercing a whole array and that array is empty, the resolved array index is 0, so the validation will error. Since the resolved index is only used when array indexing references exist, it is an optional field for ResultRef so we should change it to a pointer.

/kind bug

Signed-off-by: Yongxuan Zhang yongxuanzhang@google.com

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: examples/v1/pipelineruns/beta/7392-regression.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+apiVersion: tekton.dev/v1
```
Comment (JeromeJu) [inline examples/v1/pipelineruns/beta/7392-regression.yaml:1]: Just double-checking this is created to be run in beta integration tests to make sure future changes will not introduce regression bugs as reporeted in #7392?

Comment (Yongxuanzhang) [reply examples/v1/pipelineruns/beta/7392-regression.yaml:1]: Yes, since array results is still beta

File: pkg/apis/pipeline/v1/resultref_test.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ import (
 )
 
 func TestNewResultReference(t *testing.T) {
+	idx := 1
```
Comment (QuanZhang-William) [inline pkg/apis/pipeline/v1/resultref_test.go:31]: Does this really covers the change? IIUC we need a test case that references an empty array with [*] and test the `ResultIndex` is nil?

File: pkg/apis/pipeline/v1beta1/resultref.go [hunk 3]
```diff
@@ -159,36 +159,38 @@ func stripVarSubExpression(expression string) string {
...
+// - Output: "myTask", "aStringResult", nil, "", nil
 // Valid Example 2:
 // - Input: tasks.myTask.results.anObjectResult.key1
-// - Output: "myTask", "anObjectResult", 0, "key1", nil
+// - Output: "myTask", "anObjectResult", nil, "key1", nil
 // Valid Example 3:
 // - Input: tasks.myTask.results.anArrayResult[1]
 // - Output: "myTask", "anArrayResult", 1, "", nil
+// Valid Example 4:
+// - Input: tasks.myTask.results.Result[*]
```
Comment (QuanZhang-William) [inline pkg/apis/pipeline/v1beta1/resultref.go:170]: same as above

File: pkg/apis/pipeline/v1/resultref_test.go (hunk 2 above)
Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1/resultref_test.go:31]: Oh thanks, I think I forgot to add a test case here
